### PR TITLE
Add Kafka To BigQuery Template

### PR DESCRIPTION
**Pull request by Dippatel98**
Kafka to BigQuery template

Authentication

1. Only support SASL_PLAIN and read the username/password from the Google Cloud Secret Manager.

Input

1. Reading Kafka Avro messages (Confluent wire format) using schema registry. Supports multiple schemas in a single topic.
2. Reading Kafka Avro messages (Confluent wire format) when a static schema is provided. Schema ID is hard coded to 1.
3. Reading JSON messages.

Output

1. Writing AVRO messages to the BigQuery. If using schema registry and Confluent wire format, it can write to multiple or single BQ table/s. 
2. Writing JSON message to the BigQuery.

TODO - will be in the future PRs

1. DLQ for Kafka Sink.
2. IT and load tests for the template.
3. Reading Kafka Avro messages encoded in Single-object encoding etc mentioned at https://gist.github.com/davideicardi/e8c5a69b98e2a0f18867b637069d03a9#avro-serialization
4. Other authentication protocols such as SSL and secret managers such as HashiCorp for retrieving credentials for Kafka auth.
5. Add README

### CODE REVIEW

File: cicd/internal/workflows/maven-workflows.go [hunk 1]
```diff
@@ -117,15 +117,15 @@ func (*mvnFlags) FailAtTheEnd() string {
 }
 
 func (*mvnFlags) RunIntegrationTests() string {
-	return "-PtemplatesIntegrationTests,splunkDeps"
+	return "-PtemplatesIntegrationTests,splunkDeps,confluent-repo"
```
Comment (an2x) [inline cicd/internal/workflows/maven-workflows.go:120]: I think GH workflow changes have to be in a separate PR and merged before this one (if they're required for other changes in this PR to work).

File: v2/kafka-common/src/main/java/com/google/cloud/teleport/v2/kafka/coders/GenericRecordCoder.java [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+import org.apache.beam.sdk.coders.StringUtf8Coder;
+import org.apache.beam.sdk.extensions.avro.coders.AvroCoder;
+
+/**
+ * An {@link AtomicCoder} for {@link GenericRecord}.
+ *
+ * <p>This coder is used when the schema of the incoming {@link GenericRecord} can only be known at
+ * runtime.
+ */
+public class GenericRecordCoder extends AtomicCoder<GenericRecord> {
```
Comment (an2x) [inline v2/kafka-common/src/main/java/com/google/cloud/teleport/v2/kafka/coders/GenericRecordCoder.java:36]: This seems like the same class as inhttps://github.com/GoogleCloudPlatform/DataflowTemplates/pull/1448/files#diff-6baa0b4edc19ab903f81821ba100085478a34a74323aba6bd6e66899dba6190c.
But in #1448 it's in the common module, not kafka-common. I think common module is a better place because this coder is not kafka-specific.
OK to refactor this later in a separate PR, so just FYI.

File: v2/kafka-to-bigquery/src/main/java/com/google/cloud/teleport/v2/options/KafkaToBQFlexOptions.java [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+
+import com.google.cloud.teleport.metadata.TemplateParameter;
+import com.google.cloud.teleport.v2.kafka.options.KafkaReadOptions;
+import org.apache.beam.sdk.options.Default;
+
+/**
+ * The {@link KafkaToBQFlexOptions} class provides the custom execution options passed by the
+ * executor at the command-line.
+ */
+public interface KafkaToBQFlexOptions extends KafkaReadOptions, BigQueryStorageApiStreamingOptions {
```
Comment (an2x) [inline v2/kafka-to-bigquery/src/main/java/com/google/cloud/teleport/v2/options/KafkaToBQFlexOptions.java:26]: I'd call it KafkaToBigQueryFlexOptions (we usually don't shorten BigQuery to BQ, and even if shortened, it should be Bq instead of BQ as per the camel case style).

File: v2/kafka-to-bigquery/src/main/java/com/google/cloud/teleport/v2/transforms/AvroDynamicTransform.java [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+import org.apache.beam.sdk.transforms.DoFn.ProcessElement;
+import org.apache.beam.sdk.transforms.DoFn.Setup;
+import org.apache.beam.sdk.transforms.PTransform;
+import org.apache.beam.sdk.transforms.ParDo;
+import org.apache.beam.sdk.values.KV;
+import org.apache.beam.sdk.values.PCollection;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class AvroDynamicTransform
```
Comment (an2x) [inline v2/kafka-to-bigquery/src/main/java/com/google/cloud/teleport/v2/transforms/AvroDynamicTransform.java:56]: Please add a javadoc explaining for this class is for.

File: v2/kafka-common/src/main/java/com/google/cloud/teleport/v2/kafka/transforms/NonWireAvroDeserializer.java [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+import org.apache.avro.Schema;
+import org.apache.avro.generic.GenericDatumReader;
+import org.apache.avro.generic.GenericRecord;
+import org.apache.avro.io.DatumReader;
+import org.apache.avro.io.Decoder;
+import org.apache.avro.io.DecoderFactory;
+import org.apache.kafka.common.errors.SerializationException;
+import org.apache.kafka.common.header.Headers;
+
+public class NonWireAvroDeserializer extends KafkaAvroDeserializer {
```
Comment (an2x) [inline v2/kafka-common/src/main/java/com/google/cloud/teleport/v2/kafka/transforms/NonWireAvroDeserializer.java:29]: nit: The class name doesn't make sense to me. What's "non-wire" deserializer?

File: v2/kafka-to-bigquery/src/main/java/com/google/cloud/teleport/v2/templates/KafkaToBigQueryFlex.java [hunk 6]
```diff
@@ -0,0 +1,451 @@
...
+
+  public static PipelineResult runAvroPipeline(
+      Pipeline pipeline,
+      KafkaToBQFlexOptions options,
+      List<String> topicsList,
+      String bootstrapServers,
+      Map<String, Object> kafkaConfig)
+      throws IOException, RestClientException {
+
+    if (options.getAvroFormat().equals("NON_WIRE_FORMAT") && options.getAvroSchemaPath() == null) {
```
Comment (an2x) [inline v2/kafka-to-bigquery/src/main/java/com/google/cloud/teleport/v2/templates/KafkaToBigQueryFlex.java:224]: Can be fixed later during refactoring: this source format should share the same enum with KafkaToGcs template.

File: v2/kafka-to-bigquery/src/main/java/com/google/cloud/teleport/v2/templates/KafkaToBigQueryFlex.java [hunk 7]
```diff
@@ -0,0 +1,451 @@
...
+            + "\';");
+    return properties.buildOrThrow();
+  }
+
+  static class ThrowErrorFn<T, W> extends DoFn<FailsafeElement<T, W>, FailsafeElement<T, W>> {
+
+    @ProcessElement
+    public void processElement(ProcessContext context) {
+      FailsafeElement<T, W> element = context.element();
+      LOG.error(element.toString() + element.getErrorMessage() + element.getStacktrace());
```
Comment (an2x) [inline v2/kafka-to-bigquery/src/main/java/com/google/cloud/teleport/v2/templates/KafkaToBigQueryFlex.java:447]: OK to keep for now, but we can't have this in the production version of the template (can't log every error, this may overload the Cloud Logging API if there're too many errors). Maybe at least add a TODO to fix this later?

File: v2/kafka-to-bigquery/src/main/java/com/google/cloud/teleport/v2/transforms/AvroTransform.java [hunk 8]
```diff
@@ -0,0 +1,187 @@
...
+import org.apache.beam.sdk.transforms.DoFn;
+import org.apache.beam.sdk.transforms.DoFn.ProcessContext;
+import org.apache.beam.sdk.transforms.DoFn.ProcessElement;
+import org.apache.beam.sdk.transforms.PTransform;
+import org.apache.beam.sdk.transforms.ParDo;
+import org.apache.beam.sdk.values.PCollection;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class AvroTransform
```
Comment (an2x) [inline v2/kafka-to-bigquery/src/main/java/com/google/cloud/teleport/v2/transforms/AvroTransform.java:56]: Please add javadoc.

File: v2/kafka-to-bigquery/src/main/java/com/google/cloud/teleport/v2/transforms/BigQueryDynamicDestination.java [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+    String bqQualifiedFullName = element.getSchema().getFullName().replace(".", "-");
+    String tableName =
+        this.tableNamePrefix + (this.tableNamePrefix == "" ? "" : "-") + bqQualifiedFullName;
+    String tableSpec = this.projectName + ":" + this.datasetName + "." + tableName;
+    return new TableDestination(tableSpec, null);
+  }
+
+  @Override
+  public TableSchema getSchema(GenericRecord element) {
+    return BigQueryUtils.toTableSchema(AvroUtils.toBeamSchema(element.getSchema()));
```
Comment (an2x) [inline v2/kafka-to-bigquery/src/main/java/com/google/cloud/teleport/v2/transforms/BigQueryDynamicDestination.java:68]: I think when we write TableRows, this method is not required, and it can return null. Can you check if it works with nulls? If it does, that's gonna be a little more efficient (we won't do unnecessary schema conversions).

Comment (an2x) [reply v2/kafka-to-bigquery/src/main/java/com/google/cloud/teleport/v2/transforms/BigQueryDynamicDestination.java:68]: (OK to do it later, but at least add a TODO)

File: v2/kafka-to-bigquery/src/main/java/com/google/cloud/teleport/v2/utils/BigQueryAvroUtils.java [hunk 10]
```diff
@@ -0,0 +1,525 @@
+/*
```
Comment (an2x) [inline v2/kafka-to-bigquery/src/main/java/com/google/cloud/teleport/v2/utils/BigQueryAvroUtils.java:1]: If this was temporarily copied from Beam SDK to make some fixes, please make it clear in comments/javadoc.

**Review by an2x — CHANGES_REQUESTED** [review]

File: cicd/internal/workflows/maven-workflows.go (hunk 1 above)
Comment (Dippatel98) [reply cicd/internal/workflows/maven-workflows.go]: Created a separate PR: https://github.com/GoogleCloudPlatform/DataflowTemplates/pull/1486

File: v2/kafka-common/src/main/java/com/google/cloud/teleport/v2/kafka/transforms/NonWireAvroDeserializer.java (hunk 5 above)
Comment (Dippatel98) [reply v2/kafka-common/src/main/java/com/google/cloud/teleport/v2/kafka/transforms/NonWireAvroDeserializer.java:29]: This deserializer is for messages in non-wire format. How about renaming it to `ByteAvroDeserializer`?

File: v2/kafka-to-bigquery/src/main/java/com/google/cloud/teleport/v2/transforms/BigQueryDynamicDestination.java (hunk 9 above)
Comment (Dippatel98) [reply v2/kafka-to-bigquery/src/main/java/com/google/cloud/teleport/v2/transforms/BigQueryDynamicDestination.java:68]: IIUC this can be null if the tables are already there, we need the schema if we want to create tables at run time. I can test the behaviour though.

File: v2/kafka-common/src/main/java/com/google/cloud/teleport/v2/kafka/coders/GenericRecordCoder.java (hunk 2 above)
Comment (Dippatel98) [reply v2/kafka-common/src/main/java/com/google/cloud/teleport/v2/kafka/coders/GenericRecordCoder.java]: Done

File: v2/kafka-to-bigquery/src/main/java/com/google/cloud/teleport/v2/options/KafkaToBQFlexOptions.java (hunk 3 above)
Comment (Dippatel98) [reply v2/kafka-to-bigquery/src/main/java/com/google/cloud/teleport/v2/options/KafkaToBQFlexOptions.java]: Done

File: v2/kafka-to-bigquery/src/main/java/com/google/cloud/teleport/v2/transforms/AvroDynamicTransform.java (hunk 4 above)
Comment (Dippatel98) [reply v2/kafka-to-bigquery/src/main/java/com/google/cloud/teleport/v2/transforms/AvroDynamicTransform.java:63]: Done

File: v2/kafka-to-bigquery/src/main/java/com/google/cloud/teleport/v2/templates/KafkaToBigQueryFlex.java (hunk 7 above)
Comment (Dippatel98) [reply v2/kafka-to-bigquery/src/main/java/com/google/cloud/teleport/v2/templates/KafkaToBigQueryFlex.java:448]: Added a TODO

File: v2/kafka-to-bigquery/src/main/java/com/google/cloud/teleport/v2/transforms/AvroTransform.java (hunk 8 above)
Comment (Dippatel98) [reply v2/kafka-to-bigquery/src/main/java/com/google/cloud/teleport/v2/transforms/AvroTransform.java:62]: Done

File: v2/kafka-to-bigquery/src/main/java/com/google/cloud/teleport/v2/utils/BigQueryAvroUtils.java (hunk 10 above)
Comment (Dippatel98) [reply v2/kafka-to-bigquery/src/main/java/com/google/cloud/teleport/v2/utils/BigQueryAvroUtils.java:1]: Done

File: v2/kafka-to-bigquery/src/main/java/com/google/cloud/teleport/v2/transforms/BigQueryDynamicDestination.java (hunk 9 above)
Comment (Dippatel98) [reply v2/kafka-to-bigquery/src/main/java/com/google/cloud/teleport/v2/transforms/BigQueryDynamicDestination.java:69]: Can't return a null
```
java.lang.RuntimeException: java.lang.IllegalArgumentException: Unless create disposition is CREATE_NEVER, a schema must be specified, i.e. DynamicDestinations.getSchema() may not return null. However, create disposition is CREATE_IF_NEEDED, and  tableSpec: <tableSpec> returned null for destination %s
```

File: v2/kafka-common/src/main/java/com/google/cloud/teleport/v2/kafka/transforms/NonWireAvroDeserializer.java (hunk 5 above)
Comment (AnandInguva) [reply v2/kafka-common/src/main/java/com/google/cloud/teleport/v2/kafka/transforms/NonWireAvroDeserializer.java:29]: What about BinaryAvroDeserializer?

Comment (an2x) [reply v2/kafka-common/src/main/java/com/google/cloud/teleport/v2/kafka/transforms/NonWireAvroDeserializer.java:29]: I guess let's decide later during refactoring, when we agree on different format we support and Enum values.

File: v2/kafka-to-bigquery/src/main/java/com/google/cloud/teleport/v2/transforms/BigQueryDynamicDestination.java (hunk 9 above)
Comment (an2x) [reply v2/kafka-to-bigquery/src/main/java/com/google/cloud/teleport/v2/transforms/BigQueryDynamicDestination.java:69]: I think for the destination table we can use CREATE_NEVER disposition, then this can be null.
